### ISSUE TITLE
Can't compile code with namespace

### ISSUE THREAD

**Issue by martine** [angular/sickle#132]
```js
namespace foo {
  export var bar = 3;
}
```
produces

```js
var foo;
(function (foo) {
    foo.bar = 3;
})(foo || (foo = {}));
```

which has this error:

```
t/t.js:4: ERROR - The alias foo is assigned a value more than once.
})(foo || (foo = {}));
           ^

1 error(s), 0 warning(s)
```

**Comment by martine** [comment]
Fixing this is necessary to e.g. compile the TypeScript compiler with sickle, as that uses namespaces.

**Comment by martine** [comment]
Working around this in sickle is pretty hard.

**Comment by martine** [comment]
The reason this is especially hard is when you have nested exported namespaces.

```js
module foo {
	export module bar {
		export var x = 3;
	}
	export var y = bar.x;
}
```
becomes
```js
var foo;
(function (foo) {
    var bar;
    (function (bar) {
        bar.x = 3;
    })(bar = foo.bar || (foo.bar = {}));
    foo.y = bar.x;
})(foo || (foo = {}));
```

We need to rewrite that inner `bar.x` into `foo.bar.x` for Closure.

**Comment by mprobst** [comment]
Could we add an option to JSCompiler to suppress that error?

**Comment by martine** [comment]
It appears (from a glance at the code) it doesn't currently have an option for that.  I wonder if I could just rewrite the bit inside the function call to remove the multiple assignments.

E.g. rewrite the above to something like this:
```js
goog.provide('foo');
(function (foo) {
   ...
})(foo));
```

I would expect they don't want you mutating a namespace (imperatively adding to 'foo') though.  We might need to go as far as dropping that function wrapper, but that also doesn't seem *so* hard to do, especially if we need to recognize these function wrappers to rewrite the function param anyway.

**Comment by martine** [comment]
I take back my above idea.  In the TypeScript code, each different file has its own "namespace ts" block (see e.g. https://github.com/Microsoft/TypeScript/blob/master/src/compiler/core.ts ), so we can't translate all of them to a `goog.provide` since you're only allowed to `goog.provide` once.

I'll ponder this some more and comment again if I have any better ideas.

**Comment by martine** [comment]
Problematic source code looks like:

File 1:
```js
namespace ts { export var foo = 3; }
```

File 2:
```js
namespace ts { export var bar = 3; }
```

More thinking: if we always produce `goog.module` from these files, then there is no way to have a unified "ts" namespace across those modules.

Unless we:
1) concatenate the files before compiling?
2) stash `ts` on window (maybe?)
3) declare `ts` as externs?

In the case of TypeScript, I think concatenating best expresses the desired semantics.

**Comment by mprobst** [comment]
@martine for (1), wouldn't you potentially need the entire code base available to concatenate all the files declaring something in namespace X?

**Comment by martine** [comment]
Yes.  In the case of compiling e.g. TypeScript as a `ts_library`, we have all the source available.  In general code that writes "into" a shared namespace like my example above tends to be within a single project (because the namespace of TS namespaces is global, you usually don't share them between projects).

**Comment by evmar** [comment]
Our "fix" for this for now is just not use `namespace`.
